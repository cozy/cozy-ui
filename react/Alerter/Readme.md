Display notications!  
Include the `<Alerter />` component somewhere in your app, but only once. After that, you can use the static methods `Alerter.info`, `Alerter.success` and `Alerter.error` to trigger a notification.

```
<div>
  <button onClick={() => Alerter.info("hi!")}>Show alert</button>
  <Alerter />
</div>
```
### Alert with a button

The `ALerter` methods support an optionnal second parameter, which can be used to add a button to the notification, with the `buttonText` and `buttonAction` keys:

```
const triggerAlert = () => {
  Alerter.info("Accept our terms and conditions and subscribe to our newsletter", {
    buttonText: "ok",
    buttonAction: () => alert("Thanks!")
  });
};

<div>
  <button onClick={triggerAlert}>Show alert</button>
  <Alerter />
</div>
```
