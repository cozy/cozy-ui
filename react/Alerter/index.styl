@import '../stylus/ui-base/fonts'
@import '../stylus/ui-base/palette'
@import '../stylus/ui-components/alerts'

.coz-alerter
  @extends $alerts
