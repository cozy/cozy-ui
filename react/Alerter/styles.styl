@require '../../stylus/tools/mixins'
@require '../../stylus/settings/palette'
@require '../../stylus/components/alerts'
@require '../../stylus/components/button'

.c-alert
    @extend $alert

:global(.has-modal) .c-alert
    @extend $alert--modal

.c-alert-wrapper
    @extend $alert-wrapper

.c-alert-title
    @extend $alert-title

.c-alert--error
    @extend $alert--error

.c-alert--success
    @extend $alert--success

.c-alert--info
    @extend $alert--info

.c-alert--hidden
    @extend $alert--hidden

.c-btn--alert-error
    @extend $button-alert--error

.c-btn--alert-info
    @extend $button-alert--info

.c-btn--alert-success
    @extend $button-alert--success

