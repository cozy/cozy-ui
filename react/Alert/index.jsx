import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import merge from 'lodash/merge'
import MuiAlert from '@material-ui/lab/Alert'

import { makeStyles } from '../styles'
import Icon from '../Icon'
import CheckCircleIcon from '../Icons/CheckCircle'
import WarningIcon from '../Icons/Warning'
import WarningCircleIcon from '../Icons/WarningCircle'
import InfoIcon from '../Icons/Info'

const DEFAULT_ICON_SIZE = 16

const defaultIconMapping = {
  success: <Icon icon={CheckCircleIcon} />,
  warning: <Icon icon={WarningIcon} />,
  error: <Icon icon={WarningCircleIcon} />,
  info: <Icon icon={InfoIcon} />
}

const makeIcon = (icon, severity) => {
  // used to remove icon
  if (icon === false) {
    return false
  }

  return (
    icon ||
    (['primary', 'secondary'].includes(severity) && <Icon icon={InfoIcon} />) ||
    undefined
  )
}

const useStyles = makeStyles({
  message: {
    maxWidth: ({ block, iconSize }) =>
      block && `calc(100% - ${iconSize}px - 12px)`
  }
})

const Alert = forwardRef(
  (
    {
      className,
      classes,
      icon,
      severity,
      block,
      color,
      square,
      action,
      variant,
      children,
      ...props
    },
    ref
  ) => {
    // as primary and secondary doesn't exist on Mui Alert, we force success severity for those
    const madeSeverity = ['primary', 'secondary'].includes(severity)
      ? 'success'
      : severity
    const madeIcon = makeIcon(icon, severity)
    const iconSize = icon?.props?.size || DEFAULT_ICON_SIZE
    const styles = useStyles({ iconSize, block })

    return (
      <MuiAlert
        ref={ref}
        style={{ backgroundColor: color, borderRadius: square && 0 }}
        className={cx(
          className,
          `cozyStyles-${severity}-${variant}`,
          { block },
          { action: Boolean(action) }
        )}
        classes={merge(styles, classes)}
        variant={variant}
        severity={madeSeverity}
        iconMapping={defaultIconMapping}
        icon={madeIcon}
        action={action}
        {...props}
      >
        {children}
      </MuiAlert>
    )
  }
)

Alert.displayName = 'Alert'

Alert.propTypes = {
  className: PropTypes.string,
  icon: PropTypes.oneOfType([PropTypes.element, PropTypes.bool]),
  severity: PropTypes.oneOf([
    'primary',
    'secondary',
    'success',
    'error',
    'warning',
    'info'
  ]),
  block: PropTypes.bool,
  color: PropTypes.string,
  square: PropTypes.bool,
  variant: PropTypes.oneOf(['standard', 'outlined', 'filled'])
}

Alert.defaultProps = {
  severity: 'primary',
  block: false,
  square: false,
  variant: 'standard'
}

export default Alert
