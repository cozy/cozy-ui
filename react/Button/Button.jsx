import React from 'react'
import cx from 'classnames'
import PropTypes from 'prop-types'
import styles from './styles'

const Button = props => {
  const { theme, busy, disabled, className, children, onClick } = props
  return (
    <button
      aria-busy={busy}
      disabled={disabled}
      role="button"
      className={cx(
        styles['c-btn'], {
          [styles[`c-btn--${theme}`]] : theme
        },
        className)}
      onClick={onClick}
    >
      {children}
    </button>
  )
}

Button.propTypes = {
  children: PropTypes.node.isRequired,
  theme: PropTypes.string,
  busy: PropTypes.bool,
  disabled: PropTypes.bool,
  className: PropTypes.string,
  onClick: PropTypes.func
}

Button.defaultProps = {
  theme: '',
  busy: false,
  disabled: false,
  className: '',
  onClick: null
}

export default Button
