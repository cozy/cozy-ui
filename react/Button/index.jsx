import React from 'react'
import cx from 'classnames'
import PropTypes from 'prop-types'
import styles from './styles.styl'
import Icon, { iconPropType } from '../Icon'

import SpinnerIcon from '../Icons/Spinner'

const btnClass = function(options) {
  const { className, extension, size, theme, variant, round, align } = options
  return cx(
    styles['c-btn'],
    {
      [styles[`c-btn--${theme}`]]: theme,
      [styles[`c-btn--${size}`]]: size !== 'normal',
      [styles[`c-btn--${variant}`]]: variant,
      [styles[`c-btn--${extension}`]]: extension,
      [styles[`c-btn--${align}`]]: align,
      [styles[`c-btn--round`]]: round
    },
    className
  )
}

const transformToAria = tag => props => {
  const { busy, disabled, ...newProps } = props

  if (busy) {
    newProps['aria-busy'] = true
  }
  if (disabled) {
    newProps['aria-disabled'] = true
  }
  if (tag === 'button' && disabled) {
    newProps['disabled'] = true
  }

  return newProps
}

const tagToTransformProps = {
  button: transformToAria('button'),
  a: transformToAria('a')
}

const identity = x => x

const sizeToIconSize = {
  tiny: 8,
  small: 12,
  large: 18
}

const BaseButton = props => {
  const {
    children,
    icon,
    iconOnly,
    label,
    subtle,
    className,
    extension,
    round,
    size,
    theme,
    align,
    extraRight,
    tag: Tag,
    ...restProps
  } = props
  const transformProps = tagToTransformProps[Tag] || identity
  const tooltip = iconOnly ? label : null
  const iconOnlyClass = iconOnly ? 'u-visuallyhidden' : null

  return (
    <Tag
      {...transformProps(restProps)}
      className={btnClass({
        extension,
        align,
        round,
        size,
        theme,
        className,
        variant: subtle && 'subtle'
      })}
      title={tooltip}
    >
      <span>
        {Icon.isProperIcon(icon) ? (
          <Icon
            size={sizeToIconSize[size]}
            icon={icon}
            aria-hidden
            focusable="false"
          />
        ) : (
          icon
        )}
        {label && <span className={iconOnlyClass}>{label}</span>}
        {children}
        {extraRight && <span className="u-ml-auto">{extraRight}</span>}
        {restProps.busy && (
          <Icon
            size={sizeToIconSize[size]}
            icon={SpinnerIcon}
            spin
            className="u-ml-half"
            aria-hidden
            focusable="false"
          />
        )}
      </span>
    </Tag>
  )
}

const Button = props => <BaseButton {...props} />
const ButtonLink = props => <BaseButton {...props} />

export { Button, ButtonLink }

const DefaultButton = props => {
  if (!props.tag) {
    return <Button {...props} />
  } else {
    return <BaseButton {...props} />
  }
}

export default DefaultButton

// Proptypes (unfortunately, Styleguidist does not pick
// proptypes coming from a spread so we have to keep both
// proptypes in sync)
Button.propTypes = {
  /** DEPRECATED: please use label and icon */
  children: PropTypes.node,
  /** Label of the button */
  label: PropTypes.node.isRequired,
  /** Icon of the button */
  icon: PropTypes.oneOfType([PropTypes.node, iconPropType]),
  /** Displays only the icon, not the label */
  iconOnly: PropTypes.bool,
  theme: PropTypes.string,
  size: PropTypes.oneOf(['tiny', 'small', 'large', 'normal']),
  /** Spacing of the button */
  extension: PropTypes.oneOf(['narrow', 'full']),
  /** Button's label alignment */
  align: PropTypes.oneOf(['left', 'right', 'center']),
  /** Will make the button round */
  round: PropTypes.bool,
  /** Extra class */
  className: PropTypes.string,
  /** What to do on click */
  onClick: PropTypes.func,
  /** Adds an element to the right of the button */
  extraRight: PropTypes.PropTypes.node,

  // Only for Button
  /** Will display a spinner if true */
  busy: PropTypes.bool,
  /** Disables the button */
  disabled: PropTypes.bool,
  /** Type of the underlying `<button />` */
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
  /** Use the `subtle` alternative look for the Button  */
  subtle: PropTypes.bool
}

Button.defaultProps = {
  type: 'submit',
  tag: 'button',
  size: 'normal',
  align: 'center'
}

ButtonLink.defaultProps = {
  tag: 'a'
}
