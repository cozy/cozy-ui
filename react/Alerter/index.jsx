import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Portal from '../Portal'
import Alert from './Alert'

const MINIMUM_ALERT_DURATION = 2000

let currentId = 0

const createStore = () => {
  let state = []
  const listeners = []

  const emit = () => {
    listeners.forEach(listener => listener(state))
  }

  const addNotification = notification => {
    notification.id = currentId++
    state = [...state, notification]
    emit()
    return notification
  }

  const removeNotification = notification => {
    const id = notification.id
    const idx = state.findIndex(n => n.id === id)
    if (idx === -1) {
      return
    }
    const removed = state[idx]
    state = [...state.slice(0, idx), ...state.slice(idx + 1)]
    emit()
    return removed
  }

  const subscribe = listener => {
    listeners.push(listener)
  }

  const reset = () => {
    state = []
    emit()
  }

  return { addNotification, removeNotification, subscribe, reset }
}

const store = createStore()

class Alerter extends Component {
  state = {
    notifications: []
  }

  static reset() {
    store.reset()
  }

  /**
   * @param {string} message
   * @param {object} options
   * @public
   * @static
   */
  static info(msg, options) {
    return store.addNotification({ type: 'info', msg, options })
  }

  /**
   * @param {string} msg
   * @param {object} options
   * @public
   * @static
   */
  static success(msg, options) {
    return store.addNotification({ type: 'success', msg, options })
  }

  /**
   * @param {string} msg
   * @param {object} options
   * @public
   * @static
   */
  static error(msg, options) {
    return store.addNotification({ type: 'error', msg, options })
  }

  /**
   * Remove notification by id
   */
  static removeNotification(notification) {
    return store.removeNotification(notification)
  }

  componentDidMount() {
    this.setState({ mounted: true })
    store.subscribe(this.onStoreEvent)
  }

  onStoreEvent = newNotifications => {
    this.setState({
      notifications: newNotifications
    })
  }

  handleClose = id => {
    store.removeNotification(id)
  }

  render() {
    const { t, into } = this.props
    const { notifications } = this.state
    return (
      <Portal into={into}>
        {notifications.map(notif => (
          <Alert
            type={notif.type}
            key={notif.id}
            message={t ? t(notif.msg, notif.options) : notif.msg}
            onClose={() => this.handleClose(notif)}
            buttonText={notif.options && notif.options.buttonText}
            buttonAction={notif.options && notif.options.buttonAction}
            duration={notif.options && notif.options.duration}
            minDuration={MINIMUM_ALERT_DURATION}
          />
        ))}
      </Portal>
    )
  }
}

Alerter.propTypes = {
  /** A translation function. If ommited, messages are left intact */
  t: PropTypes.func,
  /** A selector to target a DOM node where alerts will be rendered */
  into: PropTypes.string
}

Alerter.defaultProps = {
  into: 'body'
}

export default Alerter
