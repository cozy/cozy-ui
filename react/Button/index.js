import Button from '@material-ui/core/Button'

export default Button
