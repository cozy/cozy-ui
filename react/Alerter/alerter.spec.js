'use strict'
/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import Alerter from './'
import Alert from './Alert'
import Button from '../Button'

describe('Alerter component', () => {
  beforeEach(() => {
    Alerter.reset()
  })
  //
  ;['info', 'success', 'error'].forEach(type => {
    it(`renders correctly an ${type} alert`, () => {
      const wrapper = shallow(<Alerter />)
      Alerter[type](`Test ${type} alert`)
      expect(
        wrapper
          .find(Alert)
          .dive()
          .getElement()
      ).toMatchSnapshot()
    })
  })

  it('should not render too much alerts', () => {
    const wrapper = shallow(<Alerter />)
    Alerter.info(`Test 1`)
    Alerter.info(`Test 2`)
    Alerter.info(`Test 3`)
    expect(wrapper.find(Alert).length).toBe(3)
  })

  it('renders button if buttonText provided', () => {
    const wrapper = shallow(<Alerter />)
    Alerter.info(`Test alert with button`, {
      buttonText: 'BTN'
    })
    expect(
      wrapper
        .find(Alert)
        .dive()
        .getElement()
    ).toMatchSnapshot()
  })

  it('handles dismiss provided to buttonAction', () => {
    const alerterWrapper = shallow(<Alerter />)
    Alerter.info(`Test alert with button`, {
      buttonText: 'BTN',
      buttonAction: dismiss => {
        console.log('DISMISS')
        dismiss()
      },
      duration: 20000
    })
    const alertWrapper = alerterWrapper.find(Alert)
    const alert = alertWrapper.dive()
    expect(alert.getElement()).toMatchSnapshot()
    alert.instance().base = document.createElement('div')
    // we have to wait the same delay as in the Alerter didMount
    setTimeout(() => {
      expect(alert.state('hidden')).toBe(false)
      alert
        .children()
        .find(Button)
        .simulate('click')
      expect(alert.state('hidden')).toBe(true)
    }, 20)
  })

  it('handles programmatic removal', () => {
    const wrapper = shallow(<Alerter />)
    const notif = Alerter.info(`Test alert with button`, {
      duration: 20000
    })
    Alerter.info(`Test another alert`, {
      duration: 20000
    })
    expect(wrapper.find(Alert).length).toBe(2)
    Alerter.removeNotification(notif)
    expect(wrapper.find(Alert).length).toBe(1)

    // Nothing should happen if we try to remove a non existing notification
    Alerter.removeNotification(notif)
    expect(wrapper.find(Alert).length).toBe(1)
  })
})
