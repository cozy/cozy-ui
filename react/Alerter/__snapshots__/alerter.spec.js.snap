// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alerter component handles dismiss provided to buttonAction 1`] = `
<div
  className="styles__c-alert___f9mWB styles__c-alert--hidden___22Mz1"
  role="alert"
>
  <div
    className="styles__c-alert-wrapper___3fLCt styles__c-alert--info___3mqAh"
  >
    <p>
      Test alert with button
    </p>
    <DefaultButton
      className="styles__c-btn--alert-info___1BO12"
      label="BTN"
      onClick={[Function]}
      size="tiny"
    />
  </div>
</div>
`;

exports[`Alerter component renders button if buttonText provided 1`] = `
<div
  className="styles__c-alert___f9mWB styles__c-alert--hidden___22Mz1"
  role="alert"
>
  <div
    className="styles__c-alert-wrapper___3fLCt styles__c-alert--info___3mqAh"
  >
    <p>
      Test alert with button
    </p>
    <DefaultButton
      className="styles__c-btn--alert-info___1BO12"
      label="BTN"
      onClick={[Function]}
      size="tiny"
    />
  </div>
</div>
`;

exports[`Alerter component renders correctly an error alert 1`] = `
<div
  className="styles__c-alert___f9mWB styles__c-alert--hidden___22Mz1"
  role="alert"
>
  <div
    className="styles__c-alert-wrapper___3fLCt styles__c-alert--error___oA-NC"
  >
    <p>
      Test error alert
    </p>
  </div>
</div>
`;

exports[`Alerter component renders correctly an info alert 1`] = `
<div
  className="styles__c-alert___f9mWB styles__c-alert--hidden___22Mz1"
  role="alert"
>
  <div
    className="styles__c-alert-wrapper___3fLCt styles__c-alert--info___3mqAh"
  >
    <p>
      Test info alert
    </p>
  </div>
</div>
`;

exports[`Alerter component renders correctly an success alert 1`] = `
<div
  className="styles__c-alert___f9mWB styles__c-alert--hidden___22Mz1"
  role="alert"
>
  <div
    className="styles__c-alert-wrapper___3fLCt styles__c-alert--success___1Cs5b"
  >
    <p>
      Test success alert
    </p>
  </div>
</div>
`;
